rust/src/csvreader.rs

```
@@ -0,0 +1,263 @@
+pub struct CsvReader {
+    /// Explicit schema for the CSV file
+    schema: Arc<Schema>,
+    /// Optional projection for which columns to load (zero-based column indices)
+    projection: Option<Vec<usize>>,
+    /// File reader
+    record_iter: StringRecordsIntoIter<BufReader<File>>,
+    /// Batch size (number of records to load each time)
+    batch_size: usize,
+}
+
+impl CsvReader {
+    /// Create a new CsvReader
+    pub fn new(
+        file: File,
+        schema: Arc<Schema>,
+        has_headers: bool,
+        batch_size: usize,
+        projection: Option<Vec<usize>>,
+    ) -> Self {
+        let csv_reader = csv::ReaderBuilder::new()
+            .has_headers(has_headers)
+            .from_reader(BufReader::new(file));
+
+        let record_iter = csv_reader.into_records();
+        CsvReader {
+            schema: schema.clone(),
+            projection,
+            record_iter,
+            batch_size,
+        }
+    }
+}
+
+macro_rules! build_primitive_array {
+    ($ROWS:expr, $COL_INDEX:expr, $TY:ty) => {{
+        let mut builder = PrimitiveArrayBuilder::<$TY>::new($ROWS.len() as i64);
+        for row_index in 0..$ROWS.len() {
+            match $ROWS[row_index].get(*$COL_INDEX) {
+                Some(s) if s.len() > 0 => builder.push(s.parse::<$TY>().unwrap()).unwrap(),
+                _ => builder.push_null().unwrap(),
+            }
+        }
+        Ok(Arc::new(builder.finish()) as ArrayRef)
+    }};
+}
+
+impl CsvReader {
+    /// Read the next batch of rows
+    pub fn next(&mut self) -> Option<Result<Arc<RecordBatch>, ArrowError>> {
+        // read a batch of rows into memory
+        let mut rows: Vec<StringRecord> = Vec::with_capacity(self.batch_size);
+        for _ in 0..self.batch_size {
+            match self.record_iter.next() {
+                Some(Ok(r)) => {
+                    rows.push(r);
+                }
+                Some(Err(_)) => {
+                    return Some(Err(ArrowError::ParseError(
```

Perhaps we should surface this error into the `ParseError`.